control-plane/grpc/src/replica/client.rs

```diff
@@ -0,0 +1,198 @@
+use crate::{
+    common::{
+        NodeFilter, NodePoolFilter, NodePoolReplicaFilter, NodeReplicaFilter, PoolFilter,
+        PoolReplicaFilter, ReplicaFilter, VolumeFilter,
+    },
+    replica::traits::ReplicaOperations,
+    replica_grpc::{
+        create_replica_reply, get_replicas_reply, get_replicas_request,
+        replica_grpc_client::ReplicaGrpcClient, share_replica_reply, CreateReplicaRequest,
+        DestroyReplicaRequest, GetReplicasRequest, ShareReplicaRequest, UnshareReplicaRequest,
+    },
+};
+use std::time::Duration;
+use tonic::transport::{Channel, Endpoint, Uri};
+
+use crate::{
+    grpc_opts::{timeout_grpc, Context},
+    replica::traits::{
+        CreateReplicaInfo, DestroyReplicaInfo, ShareReplicaInfo, UnshareReplicaInfo,
+    },
+};
+use common_lib::{
+    mbus_api::{v0::Replicas, ReplyError, TimeoutOptions},
+    types::v0::message_bus::{Filter, MessageIdVs, Replica},
+};
+use utils::DEFAULT_REQ_TIMEOUT;
+
+// RPC Replica Client
+#[derive(Clone)]
+pub struct ReplicaClient {
+    base_timeout: Duration,
+    endpoint: Endpoint,
+}
+
+impl ReplicaClient {
+    pub async fn init<O: Into<Option<TimeoutOptions>>>(addr: Option<Uri>, opts: O) -> Self {
+        let opts = opts.into();
+        let timeout = opts
+            .map(|opt| opt.base_timeout())
+            .unwrap_or_else(|| humantime::parse_duration(DEFAULT_REQ_TIMEOUT).unwrap());
+        let endpoint = tonic::transport::Endpoint::from(addr.unwrap())
+            .connect_timeout(timeout)
+            .timeout(timeout);
+        Self {
+            base_timeout: timeout,
+            endpoint,
+        }
+    }
+    pub async fn reconnect(
+        &self,
+        ctx: Option<Context>,
+        op_id: MessageIdVs,
+    ) -> ReplicaGrpcClient<Channel> {
+        let ctx_timeout = ctx.map(|ctx| ctx.timeout_opts).flatten();
+        match ctx_timeout {
+            None => {
+                let timeout = timeout_grpc(op_id, self.base_timeout);
+                let endpoint = self
+                    .endpoint
+                    .clone()
+                    .connect_timeout(timeout)
+                    .timeout(timeout);
+                ReplicaGrpcClient::connect(endpoint.clone()).await.unwrap()
```

Comment:
We shouldn't unwrap here, reconnect should return a Result as it's possible that we may fail to connect.